[AArch64] Update zero latency instructions in Neoverse scheduling tables

simonwallis2 opened the pull request:
NeoverseZeroMove was introduced for Neoverse-V2 and was added to V3 and V3AE. 
Use NeoverseZeroMove for Neoverse-V1, N2, N3 in the same way, including these instructions: 
MOV Xd|Wd, #0|XZR|WZR

For all the above Neoverse targets, the following instructions are also decoded as not utilizing the scheduling and execution resources of the machine: 
MOV Wd,Wn
MOV Xd,Xn

For Neoverse-N3 only, these instructions also have zero latency 
FMOV Dd, Dn
FMOV Sd, Sn
MOV Vd, Vn (vector)
MOV Zd.D, Zn.D
PTRUE
PFALSE

Commit:
[AArch64] Update zero latency instructions in scheduling tables for Neoverse cores

NeoverseZeroMove was introduced for Neoverse-V2 and was added to V3 and V3AE.
Use NeoverseZeroMove for Neoverse-V1, N2, N3 in the same way, including these instructions:
MOV Xd|Wd, #0|XZR|WZR

For all Neoverse targets, the following instructions are also decoded as not utilizing the scheduling and execution resources of the machine:
MOV Wd,Wn
MOV Xd,Xn

For Neoverse-N3 only, these instructions also have zero latency
FMOV Dd, Dn
FMOV Sd, Sn

Change-Id: I1a5f86e049798582d33d96ba99389e4b2ffb210e

Commit:
[AArch64] Update zero latency instructions in scheduling tables for Neoverse cores

NeoverseZeroMove was introduced for Neoverse-V2 and was added to V3 and V3AE.
Use NeoverseZeroMove for Neoverse-V1, N2, N3 in the same way, including these instructions:
MOV Xd|Wd, #0|XZR|WZR

For all Neoverse targets, the following instructions are also decoded as not utilizing the scheduling and execution resources of the machine:
MOV Wd,Wn
MOV Xd,Xn

For Neoverse-N3 only, these instructions also have zero latency
FMOV Dd, Dn
FMOV Sd, Sn

Change-Id: I955cfe3efc689bea305a708eb6d7259dced6fe04

rj-jesus commented on llvm/lib/Target/AArch64/AArch64SchedNeoverseN2.td:
```diff
@@ -882,7 +906,7 @@ def : SchedAlias<WriteFImm, N2Write_2c_1V>;
 // FP transfer, from gen to low half of vec reg
-def : InstRW<[N2Write_3c_1M0], (instrs FMOVWHr, FMOVXHr, FMOVWSr, FMOVXDr,
+def : InstRW<[N2Write_0or3c_1M0], (instrs FMOVWHr, FMOVXHr, FMOVWSr, FMOVXDr,
                                         FMOVHWr, FMOVHXr, FMOVSWr, FMOVDXr)>;
```
Nit: indentation

rj-jesus commented on llvm/lib/Target/AArch64/AArch64SchedNeoverseN2.td:
```diff
@@ -882,7 +906,7 @@ def : SchedAlias<WriteFImm, N2Write_2c_1V>;
 // FP transfer, from gen to low half of vec reg
-def : InstRW<[N2Write_3c_1M0], (instrs FMOVWHr, FMOVXHr, FMOVWSr, FMOVXDr,
+def : InstRW<[N2Write_0or3c_1M0], (instrs FMOVWHr, FMOVXHr, FMOVWSr, FMOVXDr,
                                         FMOVHWr, FMOVHXr, FMOVSWr, FMOVDXr)>;
```
This is unrelated to your patch, but since we're here: based on the SWOG, this doesn't look right for FP to GP moves?

rj-jesus replied:
Thanks, based on the SWOG it looks like the throughput may now be slightly higher than advertised? The new change still looks better than before, so up to you if you want to follow up with a fix.

rj-jesus reviewed (commented):
Hi, the changes generally look good, but I'm not sure we should be modelling zero-latency moves with zero micro-ops? AFAIU these instructions still count as a micro-op (towards decode bandwidth, for example) despite not using execution resources. If you can share any references to the contrary, that would be much appreciated.

Commit:
[AArch64] Update zero latency instructions in Neoverse scheduling tables

NeoverseZeroMove was introduced for Neoverse-V2 and was added to V3 and V3AE.
Use NeoverseZeroMove for Neoverse-V1, N2, N3 in the same way, including these instructions:
MOV Xd|Wd, #0|XZR|WZR

For all Neoverse targets, the following instructions are also decoded as not utilizing the scheduling and execution resources of the machine:
MOV Wd,Wn
MOV Xd,Xn

For Neoverse-N3 only, these instructions also have zero latency
FMOV Dd, Dn
FMOV Sd, Sn

Change-Id: Ie6b1f5c3f4d74f26bdd4c67c5e6c5acf6a8e00cc

rj-jesus commented:
> It looks like these zero-latency instructions still take up decode resources, which we don’t currently describe explicitly.
But the work done is largely register renaming, and there are no micro-ops dispatched to the individual pipelines.

I believe we implicitly model decode constraints in the "IssueWidth". My main concern with modelling these instructions with zero micro-ops is that it might trick the machine scheduler into assuming they can be scheduled freely, which AFAIU isn't true. Also, OP_RETIRED suggests these instructions do count as a micro-op.

Do you have any performance data that suggests this is preferable for performance or at least neutral?

rj-jesus commented:
Consider [this example](https://godbolt.org/z/hsMqeYh4n), which issues 20 `mov x9, 0` per iteration for 10^9 iterations. What I see:
```
$ perf stat -e cycles,instructions,op_retired ./a.out 

 Performance counter stats for './a.out':

     3,503,570,565      cycles                                                                
    22,003,024,649      instructions                     #    6.28  insn per cycle            
    22,003,034,412      op_retired                                                            

       1.133565071 seconds time elapsed

       1.050723000 seconds user
       0.003995000 seconds sys
```
If you normalise the IPC perf computes by 21/22 (because the CMP+B is fused), you get exactly 6.
LLVM-MCA currently works out something similar: https://godbolt.org/z/7az3YEEY9.

If we make zero-latency moves zero micro-ops, then we'll have:
```
Iterations:        100000
Instructions:      2000000
Total Cycles:      6251
Total uOps:        0

Dispatch Width:    6
uOps Per Cycle:    0.00
IPC:               319.95
Block RThroughput: 0.0
```
(The IPC becomes bottlenecked on the MicroOpBufferSize.)

It does seem to me that these instructions should be modelled as one micro-op... Unless I'm missing something or unless there's a compelling reason for us to make this change, I believe it would be better if we left it as is. What do you think?

simonwallis2 commented:
> It does seem to me that these instructions should be modelled as one micro-op... Unless I'm missing something or unless there's a compelling reason for us to make this change, I believe it would be better if we left it as is. What do you think?

I see that if making this zero micro-op change in isolation leads to llvm-mca reporting unrealistic cycle counts then that would be unhelpful. We still do not model all of the dispatch constraints described in the SWOG Special Considerations section 4.1. Modelling the MOPS per cycle limit would be key to reporting sensible numbers for an instruction with zero uOPs.

Thanks for supplying these examples, very useful. I tried modifying them to clarify to myself how the IPC is affected by MOPS limits and uOPs limits but was not able to provide a definite answer.

I propose to update the patch and remove the zero-micro-op change.
So the zero latency instructions stay as one micro-op, but we still proliferate zero latency instructions to those targets where they are not currently modelled.

Commit:
[AArch64] Update zero latency instructions in Neoverse scheduling tables

NeoverseZeroMove was introduced for Neoverse-V2 and was added to V3 and V3AE.
Use NeoverseZeroMove for Neoverse-V1, N2, N3 in the same way, including these instructions:
MOV Xd|Wd, #0|XZR|WZR

For all Neoverse targets, the following instructions are also decoded as not utilizing the scheduling and execution resources of the machine:
MOV Wd,Wn
MOV Xd,Xn

For Neoverse-N3 only, these instructions also have zero latency
FMOV Dd, Dn
FMOV Sd, Sn

Change-Id: I7a6a971cf75c60d8f75b210f0529c4ad813775a3

Commit:
[AArch64] Update zero latency instructions in Neoverse scheduling tables

NeoverseZeroMove was introduced for Neoverse-V2 and was added to V3 and V3AE.
Use NeoverseZeroMove for Neoverse-V1, N2, N3 in the same way, including these instructions:
MOV Xd|Wd, #0|XZR|WZR

For all Neoverse targets, the following instructions are also decoded as not utilizing the scheduling and execution resources of the machine:
MOV Wd,Wn
MOV Xd,Xn

For Neoverse-N3 only, these instructions also have zero latency
FMOV Dd, Dn
FMOV Sd, Sn

Change-Id: I95c53d373f35bb0bea5174a16c7ab3ac25acf684

simonwallis2 commented:
I reverted my changes for V2, V3, V3A and their corresponding tests,
restoring the old behaviour of zero latency not implying zero micro-ops,

I updated the changes and tests for N2, N3, V1 so they match V2,V3,V3A.
I made it explicit in N3.td and V1.td that NumMicroOps is 1.
Note that for N3, this removes the zero-micro-ops count from the instruction setffr (only).

rj-jesus commented on llvm/lib/Target/AArch64/AArch64SchedNeoverseN3.td:
```diff
@@ -75,7 +75,7 @@ def : WriteRes<WriteHint,    []> { let Latency = 1; }
 def N3Write_0c : SchedWriteRes<[]> {
     let Latency = 0;
-    let NumMicroOps = 0;
+    let NumMicroOps = 1;
```
You could remove this if you want, but up to you.

simonwallis2 commented on llvm/lib/Target/AArch64/AArch64SchedNeoverseN3.td:
```diff
@@ -855,10 +873,11 @@ def : SchedAlias<WriteFCvt, N3Write_3c_1V0>;
 // FP move, register
-def : InstRW<[N3Write_2c_1V], (instrs FMOVHr, FMOVSr, FMOVDr)>;
+def : InstRW<[N3Write_2c_1V], (instrs FMOVHr)>;
+def : InstRW<[N3Write_0c], (instrs FMOVSr, FMOVDr)>;
```
Sure. I have added these MOV instructions, all PFALSE, and the ALL variants of PTRUE and PTRUES.

Commit:
[AArch64] Update zero latency instructions in Neoverse scheduling tables

NeoverseZeroMove was introduced for Neoverse-V2 and was added to V3 and V3AE.
Use NeoverseZeroMove for Neoverse-V1, N2, N3 in the same way, including these instructions:
MOV Xd|Wd, #0|XZR|WZR

For all Neoverse targets, the following instructions are also decoded as not utilizing the scheduling and execution resources of the machine:
MOV Wd,Wn
MOV Xd,Xn

For Neoverse-N3 only, these instructions also have zero latency
FMOV Dd, Dn
FMOV Sd, Sn

Change-Id: Ibbc0ba1da02dd4bf5ca28b33164d8fa4e93958d6

rj-jesus commented on llvm/lib/Target/AArch64/AArch64SchedNeoverseN3.td:
```diff
@@ -1564,10 +1590,11 @@ def : InstRW<[N3Write_2c_1M], (instregex "^REV_PP_[BHSD]")>;
 
 // Predicate set/initialize, set flags
-def : InstRW<[N3Write_2c_1M], (instregex "^PTRUES_[BHSD]")>;
+def : InstRW<[N3Write_0or2c_1M], (instregex "^PTRUES_[BHSD]")>;
```
PTRUES isn't listed in Section 4.11, but it is mentioned in Table 2-23, so I'll assume the latter to be correct (matching what you've implemented).

rj-jesus commented on llvm/lib/Target/AArch64/AArch64SchedNeoverseN3.td:
```diff
@@ -1564,10 +1590,11 @@ def : InstRW<[N3Write_2c_1M], (instregex "^REV_PP_[BHSD]")>;
 
 // Predicate set
-def : InstRW<[N3Write_2c_1M], (instregex "^PFALSE", "^PTRUE_[BHSD]")>;
+def : InstRW<[N3Write_0c], (instregex "^PFALSE")>;
```
You could also give the opcode directly here.

rj-jesus commented on llvm/lib/Target/AArch64/AArch64SchedPredNeoverse.td:
```diff
@@ -80,5 +80,22 @@ def NeoverseZeroMove : MCSchedPredicate<
+                                     CheckSameRegOperand<1, 2>]>,
+                         ]>>;
+
+def NeoverseAllElement : MCSchedPredicate<
```
Any objections to naming this `NeoverseAllActivePredicate`? We have similarly named methods in ISel and TTI to check for all-active predicates.

rj-jesus reviewed (commented):
A few more suggestions but otherwise this is almost good to go.
Thanks for the patience. :)

Commit:
[AArch64] Update zero latency instructions in Neoverse scheduling tables

NeoverseZeroMove was introduced for Neoverse-V2 and was added to V3 and V3AE.
Use NeoverseZeroMove for Neoverse-V1, N2, N3 in the same way, including these instructions:
MOV Xd|Wd, #0|XZR|WZR

For all Neoverse targets, the following instructions are also decoded as not utilizing the scheduling and execution resources of the machine:
MOV Wd,Wn
MOV Xd,Xn

For Neoverse-N3 only, these instructions also have zero latency
FMOV Dd, Dn
FMOV Sd, Sn

Change-Id: I2d51b0ee6736d14f8212583f234431c555cc2574

rj-jesus reviewed (approved):
LGTM, cheers. Just please fix the conflict in `llvm/lib/Target/AArch64/AArch64SchedNeoverseN3.td` and let the CI tests run. :)

Commit:
Merge branch 'main' into sched-zero-latency
